rviz plugin: joints tab

rhaschke opened the pull request:
I was missing for a long time the possibility to directly actuate joints in the rviz motion planning plugin.
This PR adds a new tab "Joints" to the panel, which lists all joints of the current move group and allows to operate them via sliders within their bounds. For unbounded joints, an editing field will be displayed.
Additionally to this, sliders are displayed to navigate the redundant space of the robot. If there is only a single redundant DoF, the approach works quite nice already (see the Panda video).
For multiple redundant DoFs, we need some more work to nicely align nullspace directions between different frames.
![redundancy-exploration](https://user-images.githubusercontent.com/5376030/51172208-08f50a00-18b3-11e9-8df5-26c0eb7f2c6c.gif)

This PR also introduces methods `harmonizePositions()` in `JointModel` and `RobotState` that map joint angles of revolute joints that are outside the boundaries back into the feasible range by adding / subtracting multiples of 2pi. Such a function was implemented in several IK plugins before - I have seen it in LMA and IKFast...

simonschmeisser commented:
Looks nice! The redundancy slider is not completely obvious to me however. Maybe move it below the joints list with a clear visual separation and a header? Will this slider also appear in singularities?

rhaschke commented:
> Maybe move it below the joints list with a clear visual separation and a header?

To emphasize the different semantics of the NS sliders, I deliberately placed them next to the joint sliders and not below. If there are many joints, the NS sliders wouldn't be visible as well.

> Will this slider also appear in singularities?

Yes, of course. As an example, have a look at the Panda arm in zero pose, which has two nullspace dimensions (because three joint axes are aligned).

mlautman commented on moveit_core/robot_model/src/joint_model.cpp:
```diff
@@ -90,6 +90,31 @@ int JointModel::getLocalVariableIndex(const std::string& variable) const
 
+bool JointModel::harmonizePosition(double* values, const Bounds& other_bounds) const
+{
+  if (type_ != REVOLUTE)
```
see comment about moving this to RevoluteJointModel

mlautman commented on moveit_core/robot_model/src/joint_model.cpp:
```diff
@@ -90,6 +90,31 @@ int JointModel::getLocalVariableIndex(const std::string& variable) const
+    return false;
+
+  bool modified = false;
+  if (*values < other_bounds[0].min_position_)
```
what if the joint is continuous?

rhaschke replied:
Which behaviour would you expect for continuous joints?

rhaschke replied:
Continuous joints have their "bounds" between -Pi and Pi. Hence `harmonizePosition` will constrain their joint position to this range as well.

mlautman commented on moveit_ros/visualization/motion_planning_rviz_plugin/src/motion_planning_frame_joints.cpp:
```diff
@@ -0,0 +1,535 @@
+          break;
+        case ProgressBarDelegate::PercentageRole:
+          if (const moveit::core::VariableBounds* bounds = getVariableBounds(index))
+            return static_cast<int>((value - bounds->min_position_) / (bounds->max_position_ - bounds->min_position_) *
```
Does this correctly handle continuous joints?

rhaschke replied:
Didn't yet worked. Should be fixed now.

rhaschke commented on moveit_core/robot_model/include/moveit/robot_model/joint_model.h:
```diff
@@ -297,6 +297,14 @@ class JointModel
       Return true if changes were made. */
   virtual bool enforcePositionBounds(double* values, const Bounds& other_bounds) const = 0;
 
+  /** Harmonize position of revolute joints, adding/subtracting multiples of 2*Pi to bring them back into bounds.
```
That's also an option.

rhaschke replied:
Done.

rhaschke commented on moveit_core/robot_state/include/moveit/robot_state/robot_state.h:
```diff
@@ -1511,6 +1511,15 @@ as the new values that correspond to the group */
       updateMimicJoint(joint);
     }
   }
+  void harmonizePositions();
```
Isn't that as obvious as for enforceBounds?

davetcoleman commented:
This is a great feature, well done!

> Maybe move it below the joints list with a clear visual separation and a header?

I agree with @simonschmeisser the current UI layout isn't the most intuitive. We need some kind of label or icon indicating what the vertical slider does and how its different than the other sliders. I like his idea of having it below.

Also, the highlighting of a particular joint row is confusing to me - can we disable that behavior? In the gif it mislead my initial understanding of what the redundant slider does.

Not necessary in this PR but just a future suggestion: in the original arm_navigation project we had circular sliders overlayed *on each joint* that allowed this same per-joint functionality. It was pretty slick and was basically a simplified interactive marker:
![image](https://user-images.githubusercontent.com/561060/51419788-0d832200-1b42-11e9-8837-9a38ad4a9ec0.png)

gavanderhoorn commented:
@davetcoleman wrote:
> Not necessary in this PR but just a future suggestion: in the original arm_navigation project we had circular sliders overlayed _on each joint_ that allowed this same per-joint functionality. It was pretty slick and was basically a simplified interactive marker:
> ![image](https://user-images.githubusercontent.com/561060/51419788-0d832200-1b42-11e9-8837-9a38ad4a9ec0.png)

Something like this: [awesomebytes/move_joints_interactive](https://github.com/awesomebytes/move_joints_interactive)?

It's not perfect, but is the closest to how `arm_navigation` did it I've found so far.

rhaschke commented:
> I like his idea of having the sliders below.

Ok, moved them below and added a label now.

> The highlighting of a particular joint row is confusing to me - can we disable that behavior?

This is the standard highlighting of the selected joint row (which stays highlighted even if focus moves somewhere else). @davetcoleman, what kind of behaviour do you want to have?

> ... in the original arm_navigation project we had circular sliders overlayed on each joint ...

I like the idea to have joint sliders, but only if there is no IK solver available / no Cartesian marker.
Otherwise the multitude of markers will become overwhelming / confusing too. Having Cartesian markers on all end-effectors (finger tips) of our bimanual Shadow hand setup as well as two markers on the arm end-effectors, we already sum up to 14 markers:
![bimanual](https://user-images.githubusercontent.com/5376030/51460996-3c291480-1d5d-11e9-8b52-1a34ae9f49cb.png)

Also, choosing the right size and position of the joint markers is tricky...

davetcoleman commented:
> Something like this: awesomebytes/move_joints_interactive?

Yes exactly. I think this behavior should be an optional view that you wouldn't want at the same time as the normal goal pose interactive marker. 

This whole interface @rhaschke is adding could likely be in its own Display panel, rather than inside the moveit one. What do you think about that @rhaschke? Or are you using a lot of IK dependencies that would be bulky to load in a different Rviz plugin?

rhaschke commented:
> It concerns me we now need scroll bars to access all the tabs in the motion planning plugin

Not at my side. Seems like you use a wider system font...
I wouldn't care too much to have scroll bars though.

> The slider interface is very hard to use. Why did you choose this non-standard GUI approach?

In first place this is a TableView and not a gui interaction. As I said, primarily I wanted to monitor the joint values (instead of changing them).

> The nullspace exploration bar would make you think it represents the min and max range of the free joint.

There is no min and max written. Why would you think so? By the way, there is nothing like a "free joint", nor "nullspace coordinates". There is only a non-linear manifold which you can navigate.

> In fact all it really is doing is providing a left and right button without any known min or max. I think two buttons with arrows would be more intuitive e.g. < >

No, that's not true. It's a velocity control slider: The more far to the left/right you move, the faster the robot moves in the nullspace. That's not possible with simple buttons.

rhaschke commented on moveit_core/robot_model/include/moveit/robot_model/revolute_joint_model.h:
```diff
@@ -58,6 +58,7 @@ class RevoluteJointModel : public JointModel
                                         const double distance) const override;
   bool enforcePositionBounds(double* values, const Bounds& other_bounds) const override;
   bool satisfiesPositionBounds(const double* values, const Bounds& other_bounds, double margin) const override;
+  bool harmonizePosition(double* values, const Bounds& other_bounds) const override;
```
It's explained in the base class.

rhaschke commented on moveit_ros/visualization/motion_planning_rviz_plugin/include/moveit/motion_planning_rviz_plugin/motion_planning_frame_joints.h:
```diff
@@ -0,0 +1,236 @@
+
+/* Author: Robert Haschke */
+
+#pragma once
```
It's way simpler to write and to stay consistent with names...

davetcoleman replied:
Certainly its simpler!

rhaschke commented on moveit_ros/visualization/motion_planning_rviz_plugin/include/moveit/motion_planning_rviz_plugin/motion_planning_frame_joints.h:
```diff
@@ -0,0 +1,236 @@
+{
+  Q_OBJECT
+  moveit::core::RobotState robot_state_;
+  const moveit::core::JointModelGroup* jmg_;
```
Are you joking?

davetcoleman replied:
Check out the other moveit code. Also, Google style:

> The public section should be first, followed by the protected and finally the private section.

rhaschke commented on moveit_ros/visualization/motion_planning_rviz_plugin/src/motion_planning_frame_joints.cpp:
```diff
@@ -0,0 +1,537 @@
+{
+  ui_->joints_view_->setModel(model);
+  ui_->joints_view_label_->setText(
+      QString("Group joints of %1 state").arg(model == start_state_model_.get() ? "start" : "goal"));
```
As said, I see this primarily as a display and not as an adjustment tool.
Why do you want to change case of all worlds?

davetcoleman replied:
When I ran your code locally it looked sloppy without capitalization, as I see this as a section title

davetcoleman replied:
Also, you presented this primarily as a adjustment tool in the description of this PR:

> I was missing for a long time the possibility to directly actuate joints in the rviz motion planning plugin.

rhaschke replied:
Touché. But, actually I started from a display and it turned into an interaction later...
No matter. I don't like to switch to "normal" sliders for the following reasons:
- Continuous or unbounded joints cannot be handled with a slider in a meaningful way. For them, an input field is necessary.
- Using a model-view approach is much more flexible here.

I could have a look into how to improve the interaction though. I remember there are several modes available for a view to enter the editor of an item. But that's nothing I can promise anytime soon. MoveIt is already consuming way to much of my time and there definitely more pressing issues in MoveIt.

simonschmeisser replied:
I'd prefer sliders as well. It should be possible to have sliders as Delegate in a TableView as well

But either one would be an improvement and as you said it's no super critical feature

rhaschke replied:
Having a slider as a delegate editor should be possible as well. However, then the display changes from progress bar (in display mode) to slider (in edit mode) which is maybe strange as well.
This doesn't change the issue, that the editor in the view needs to be triggered by a double click.
As far as I know, a slider doesn't come with a label automatically. As there is only a single widget allowed as an editor, more effort is needed to create a wrapper widget combining a slider and a label in this case. If you volunteer to contribute this, you are very welcome to do so.

davetcoleman replied:
Why not just have a QT Group (perhaps encapsulated as a widget) with horizontal layout with a label widget and traditional slider widget? The slider could still reflect the status of the joints via update calls to adjust them automatically.

rhaschke replied:
That's exactly what I suggested, although there is nothing like a Qt Group. Do you mean [QGroupBox](http://doc.qt.io/qt-5/qgroupbox.html)? I guess, we don't want to have it's frame.

rhaschke replied:
I agree that the look-and-feel is not yet optimal. But, just to clarify the situation: If we don't drop the idea of the model-view approach (which I will not do, somebody else is cordially invited to do so though), we cannot show the sliders _all the time_. If editing is triggered, the display will change from a progress bar to a slider (+ label), which I think is strange too.

davetcoleman replied:
Yes I think QGroupBox is what I meant, from the top of my head. I'm fairly certain you can disable the frame via the properties as I've done that before.

I'm not sure why you are saying the model-view approach is preventing you from having a better UI. Why could there not be a callback to auto-adjust the sliders as the robot moves, which is disabled if the user clicks on the sliders?

rhaschke replied:
QGroupBox is intended to group buttons, not arbitrary widgets, but yeah might work for this purpose.

The model-view approach distinguishes between display and editing. Usually, there is only one editor (slider) open in a view at any time. If that's fine for you, great. If you want to see all the sliders all the time, we need a completely different approach, namely creating all those slider widgets as you do, e.g. in setup assistant.

rhaschke commented on moveit_ros/visualization/motion_planning_rviz_plugin/src/motion_planning_frame_joints.cpp:
```diff
@@ -0,0 +1,537 @@
+/*********************************************************************
```
github does that so often. All these classes are rather small and often the copyright disclaimer is longer than the actual source code. That's why, I decided to keep everything together.

davetcoleman commented:
> No, that's not true. It's a velocity control slider: The more far to the left/right you move, the faster the robot moves in the nullspace. That's not possible with simple buttons.

I observed it closer and I've confirmed it does move faster, my bad. It can easily be missed, though.

I think this is a great contribution and don't want to hold it up too long from being merged. However, unlike more internal code cleanliness and layout, its much less desirable to be iterating on user-facing UIs and I'd like to get that right now. From a product manager perspective, I'd like to see the slider interactions cleaned up before this gets merged

rhaschke commented:
I tweaked the view settings a bit to get rid of the double-click experience (f4d33d2). However, mouse interaction still feels strange, as the progress-bar sliders reacts w/o mouse button being pressed and there is another click+release needed to finish the editor.
I also quickly added a QSlider as an editor (4ea2c87). Here usage is even worse: you cannot "leave" the editor other than selecting something else.
Maybe indeed, we need to drop the idea of using the model-view approach but use classic widgets instead. I don't have cycles for this though and everybody taking over is very welcome.

davetcoleman commented:
I re-ran this locally and I appreciate the improved UI with slider widget. However, the robot doesn't move until the mouse looses focus from the slider bar section. This is still unintuitive...

davetcoleman commented:
I changed the branch to master.

felixvd commented:
I'm excited about this no matter if the sliders behave a little unintuitively. Would love to see it merged.

rhaschke commented:
I've put some more effort into this to get a more intuitive/immediate mouse behaviour.
